nixops/backends/libvirtd.py

```diff
@@ -130,62 +188,99 @@ def create(self, defn, check, allow_reboot, allow_recreate):
+
+        temp_disk_path = os.path.join(self.depl.tempdir, 'disk-{}.qcow2'.format(self.name))
+        shutil.copyfile(os.path.join(temp_image_path, 'disk.qcow2'), temp_disk_path)
+        # Rebase onto empty backing file to prevent breaking the disk image
+        # when the backing file gets garbage collected.
+        self._logged_exec(["qemu-img", "rebase", "-f", "qcow2", "-b",
+                           "", temp_disk_path])
+
+        self.logger.log("uploading disk image...")
+        image_info = self._get_image_info(temp_disk_path)
+        self._vol = self._create_volume(image_info['virtual-size'], image_info['actual-size'])
+        self._upload_volume(temp_disk_path, image_info['actual-size'])
+
+    def _get_image_info(self, filename):
+        output = self._logged_exec(["qemu-img", "info", "--output", "json", filename], capture_stdout=True)
+        return json.loads(output)
+
+    def _create_volume(self, virtual_size, actual_size):
+        xml = '''
+        <volume>
+          <name>{name}</name>
+          <capacity>{virtual_size}</capacity>
+          <allocation>{actual_size}</allocation>
+          <target>
+            <format type="qcow2"/>
+          </target>
+        </volume>
+        '''.format(
+            name="{}.qcow2".format(self._vm_id()),
+            virtual_size=virtual_size,
+            actual_size=actual_size,
+        )
+        vol = self.pool.createXML(xml)
+        self._vol = vol
+        return vol
 
-    def _make_domain_xml(self, defn):
-        qemu_executable = "qemu-system-x86_64"
-        qemu = spawn.find_executable(qemu_executable)
-        assert qemu is not None, "{} executable not found. Please install QEMU first.".format(qemu_executable)
+    def _upload_volume(self, filename, actual_size):
+        stream = self.conn.newStream()
+        self.vol.upload(stream, offset=0, length=actual_size)
+
+        def read_file(stream, nbytes, f):
+            return f.read(nbytes)
 
-        def maybe_mac(n):
-            if n == self.primary_net:
-                return '<mac address="' + self.primary_mac + '" />'
-            else:
-                return ""
+        with open(filename, 'rb') as f:
+            stream.sendAll(read_file, f)
+            stream.finish()
+
+    def _get_qemu_executable(self):
+        domaincaps_xml = self.conn.getDomainCapabilities(
+            emulatorbin=None, arch='x86_64', machine=None, virttype='kvm',
+        )
+        domaincaps = ElementTree.fromstring(domaincaps_xml)
+        return domaincaps.find('./path').text.strip()
+
+    def _make_domain_xml(self, defn):
+        qemu = self._get_qemu_executable()
 
         def iface(n):
             return "\n".join([
-                '    <interface type="network">',
-                maybe_mac(n),
-                '      <source network="{0}"/>',
+                '    <interface type="{interface_type}">',
+                '      <source {interface_type}="{source}"/>',
                 '    </interface>',
```

Comment:
I have problems when configuring bridges like `Cannot get interface MTU on 'default': No such device`
As a quick hack as I work with bridges I kinda reverted the fix with this hybrid change:
 ```
        def iface(n):
            return "\n".join([
                '    <interface type="network">',
                '      <source network="{0}"/>',
                '    </interface>',
                ]).format(n.source)

```
which works